src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
                 _isReferenced = true;
             }
 
-            private void AddItemsToCache(ImmutableHashSet<string> globsToIgnore, ImmutableList<ItemData> items)
+            private void AddItemsToCache(ImmutableHashSet<string> globsToIgnore, OrderedItemDataCollection items)
             {
                 if (_cache == null)
                 {
-                    _cache = new Dictionary<ISet<string>, ImmutableList<ItemData>>();
+                    _cache = new Dictionary<ISet<string>, OrderedItemDataCollection>();
                 }
 
                 _cache[globsToIgnore] = items;
             }
         }
 
+        /// <summary>
+        /// A collection of ItemData that maintains insertion order and internally optimizes some access patterns, e.g. bulk removal
+        /// based on normalized item values.
+        /// </summary>
+        internal sealed class OrderedItemDataCollection
+        {
+            #region Inner types
+
+            /// <summary>
+            /// An efficient multi-value wrapper holding one or more versioned items.
+            /// </summary>
+            internal struct DictionaryValue : IEnumerable<I>
+            {
+                /// <summary>
+                /// The version of the containing collection at the time this value was last changed.
+                /// </summary>
+                private int _version;
+
+                /// <summary>
+                /// Holds one value or a list of values.
+                /// </summary>
+                private object _value;
+
+                public DictionaryValue(int version, I item)
+                {
+                    _version = version;
+                    _value = item;
+                }
+
+                public void Add(int version, I item)
+                {
+                    if (_value is List<I> list)
+                    {
+                        if (version != _version)
+                        {
+                            list = new List<I>(list);
+                        }
+                        list.Add(item);
+                    }
+                    else
+                    {
+                        list = new List<I>
+                        {
+                            (I)_value,
+                            item
+                        };
+                    }
+                    _version = version;
+                    _value = list;
+                }
+
+                public IEnumerator<I> GetEnumerator()
```

Comment:
This enumerator allocates, maybe a `struct Enumerator` can be used here to avoid this allocations on operations?

Reply:
Fixed by adding a struct Enumerator, thank you!